pkg/cloud/services/compute/instance_types.go

```diff
@@ -132,51 +137,101 @@ func (is *InstanceStatus) InstanceIdentifier() *InstanceIdentifier {
+	// addresses as in the example output here:
+	// https://docs.openstack.org/api-ref/compute/?expanded=show-server-details-detail#show-server-details
+	//
+	// Here we convert the interface{} into something more usable by
+	// marshalling it to json, then unmarshalling it back into our own
+	// struct.
+	addressesByNetwork := make(map[string][]corev1.NodeAddress)
 	for networkName, b := range is.server.Addresses {
 		list, err := json.Marshal(b)
 		if err != nil {
 			return nil, fmt.Errorf("error marshalling addresses for instance %s: %w", is.ID(), err)
 		}
-		var networkList []networkInterface
-		err = json.Unmarshal(list, &networkList)
+		var interfaceList []networkInterface
+		err = json.Unmarshal(list, &interfaceList)
 		if err != nil {
 			return nil, fmt.Errorf("error unmarshalling addresses for instance %s: %w", is.ID(), err)
 		}
 
-		addresses[networkName] = networkList
-	}
+		var addresses []corev1.NodeAddress
+		for i := range interfaceList {
+			address := &interfaceList[i]
 
-	return &InstanceNetworkStatus{addresses}, nil
-}
+			// Only consider IPv4
+			if address.Version != 4 {
+				is.logger.V(6).Info("Ignoring IPv%d address %s: only IPv4 is supported", address.Version, address.Address)
+				continue
+			}
 
-func (ns *InstanceNetworkStatus) IP() string {
-	// Return the last listed non-floating IPv4 from the last listed network
-	// This behaviour is wrong, but consistent with the previous behaviour
-	// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/4debc1fc4742e483302b0c36b16c076977bd165d/pkg/cloud/services/compute/instance.go#L973-L998
-	// XXX: Fix this
-	for _, vifs := range ns.addresses {
-		for _, vif := range vifs {
-			if vif.Version == 4.0 && vif.Type != "floating" {
-				return vif.Address
+			var addressType corev1.NodeAddressType
+			switch address.Type {
+			case "floating":
+				addressType = corev1.NodeExternalIP
+			case "fixed":
+				addressType = corev1.NodeInternalIP
+			default:
+				is.logger.V(6).Info("Ignoring address %s with unknown type '%s'", address.Address, address.Type)
+				continue
 			}
+
+			addresses = append(addresses, corev1.NodeAddress{
+				Type:    addressType,
+				Address: address.Address,
+			})
 		}
+
+		addressesByNetwork[networkName] = addresses
 	}
-	return ""
+
+	return &InstanceNetworkStatus{addressesByNetwork}, nil
+}
+
+// Addresses returns a list of NodeAddresses containing all addresses which will
+// be reported on the OpenStackMachine object.
+func (ns *InstanceNetworkStatus) Addresses() []corev1.NodeAddress {
+	// We want the returned order of addresses to be deterministic to make
+	// it easy to detect changes and avoid unnecessary updates. Iteration
+	// over maps is non-deterministic, so we explicitly iterate over the
+	// address map in lexical order of network names. This order is
+	// arbitrary.
+	networks := make([]string, 0, len(ns.addresses))
+	for network := range ns.addresses {
+		networks = append(networks, network)
+	}
```

Comment:
I read your comment, looked at my code, and agreed with you.

I was trying to work out why I wrote it this way when I realised the code is actually correct, because the range iterates over only the keys of the map, not the elements.

I think that's a fairly clear indication this code needs another comment 😂 I'll add one.

Reply:
Added a comment.